### ISSUE TITLE
`repositories` function clashes with `Project.repositories` from the Kotlin DSL (Gradle 8.1)

### ISSUE THREAD

**Issue by szpak** [gradle-nexus/publish-plugin#220]
Unrelated to the above but also related to Gradle 8 compatibility (at least with Gradle 8.1.1):

![image](https://user-images.githubusercontent.com/8314616/234118093-edb8f874-60e7-450f-9b60-7796629ae494.png)

The `repositories` function is now clashing with `Project.repositories` from the Kotlin DSL. For now it seems to still compile, but it does show an error in the IDE. Doing something like the following fixes it (since thankfully the `Project.repositories` argument has a different name):

![image](https://user-images.githubusercontent.com/8314616/234118587-bf553f01-537d-409a-94d1-f57ed7d34aa8.png)

It might make sense to introduce an alias to `repositories` such as `nexusRepositories` and perhaps deprecate the former.

If I'm missing something obvious and there's a less hacky workaround please do let me know!

I tried going over the Gradle release notes to check if this change in behaviour is mentioned anywhere but couldn't find anything to reference. I'd appreciate if someone could find a reference as to why this is now happening.

Also, please let me know if you'd prefer that I create a new issue to track this.

_Originally posted by @YarnSphere in https://github.com/gradle-nexus/publish-plugin/issues/208#issuecomment-1520845369_

**Comment by szpak** [comment]
@YarnSphere seems to [find a workaround](https://github.com/gradle-nexus/publish-plugin/issues/208#issuecomment-1529047972):

---

Regarding https://github.com/gradle-nexus/publish-plugin/issues/208#issuecomment-1520845369, where I say:

> If I'm missing something obvious and there's a less hacky workaround please do let me know!

The less hacky workaround is simply:

```kt
nexusPublishing {
    this.repositories {
        sonatype()
    }
}
```

Which I find acceptable (I don't know how I originally missed this :sweat_smile:).

---

I wonder if - in addition to mention it in the documentation - the plugin should do anything else (considering the backward compatibility)?

**Comment by TWiStErRob** [comment]
> For now it seems to still compile, but it does show an error in the IDE.

This means the regression bug is in the IDE or the Gradle v IDEA integration. Is there a repro that we can use to report it to Gradle?

**Comment by Vampire** [comment]
Seems to work just fine with Gradle 8.2 and IntelliJ IDEA 2023.1.3:
![image](https://github.com/gradle-nexus/publish-plugin/assets/325196/2b2c94b5-44b2-4a60-b6f0-7496dc21d7ee)

**Comment by Vampire** [comment]
Actually even with 8.1 it works fine here.

**Comment by TWiStErRob** [comment]
@Vampire could it be some combination of imports?

**Comment by Vampire** [comment]
Wouldn't know why or how and there is no reproducer project attached here, is there?

**Comment by Vampire** [comment]
Maybe Kotlin IntelliJ plugin version

**Comment by Vampire** [comment]
231-1.8.21-IJ9161.38 here

**Comment by YarnSphere** [comment]
This example project showcased this issue, last I checked (at the time I wrote the comment this issue references): https://github.com/YarnSphere/nexus-publish-gradle-8

The IDE would complain without the `this` in the following line: https://github.com/YarnSphere/nexus-publish-gradle-8/blob/main/build.gradle.kts#L46

**Comment by Vampire** [comment]
Interesting, with that it indeed happens

**Comment by YarnSphere** [comment]
Are you perhaps declaring your `nexusPublishing` configuration outside the scope of a `Project` (possibly outside of a `build.gradle.kts`?

**Comment by Vampire** [comment]
Yes and no, found the difference to mine.
I have it in a convention plugin, there it works perfectly fine.
If I move the block to the main build's build script, it happens the same in my project.

**Comment by Vampire** [comment]
Btw. there is also this work-around:
```kts
nexusPublishing.repositories {
    sonatype()
}
```

**Comment by Vampire** [comment]
Is this reported to somewhere already?
Seems like someone else should fix this. :-D

**Comment by YarnSphere** [comment]
I just took the error message as an implication that this is an intended consequence of the changes being made to the resolution algorithm, i.e. that this isn't a bug per se.

That's why I originally proposed the renaming/deprecation of the `repositories` function in this project. I am, however, sufficiently happy with either of the proposed work-arounds, since they're low friction enough (in fact, I prefer yours to mine :smiley:).

Do you believe this to be a bug that should be reported upstream? From the error message I assumed it's expected.

**Comment by Vampire** [comment]
Actually, I interpreted the error message the other way around.
That IJ thinks it is currently resolved to the `Project` one and soon will be changed.

Also because I wonder why it should ever be the case, that it resolves to the extensions function `fun Project.repositories` while within `nexusPublishing { ... }` `this` is the `NexusPublishingExtension` which has a member `fun repositories` as direct members should always win. Especially as the extension function is not even for that `this`, but for an outer one, or actually for the delegate of the enclosing script instance.

And even if this resolution would "change soon", I don't think the IDE should display an error there for perfectly valid and working code, that is just non-sense and user unfriendly. So even if this is intended, it should imho be reported and change immediately by JetBrains.

Btw. another work-around variant, maybe even better:
```kts
nexusPublishing {
    repositories.sonatype()
}
```

**Comment by Vampire** [comment]
Or another one:
```kts
import org.gradle.kotlin.dsl.repositories as builtInRepositories

nexusPublishing {
    repositories {
        sonatype()
    }
}
```

**Comment by marcphilipp** [comment]
This is caused by an issue in IntelliJ idea: [KTIJ-26463](https://youtrack.jetbrains.com/issue/KTIJ-26463)

**Comment by 3flex** [comment]
KTIJ issue is fixed in 2023.3, can close?

**Comment by szpak** [comment]
@YarnSphere , as an original reported, is it fixed for you in IDEA 2023.3?

**Comment by YarnSphere** [comment]
Seems fixed to me! Feel free to close! :)
